pkg/standalone-utils/contracts/test/MockTetuSmartVault.sol

```diff
@@ -0,0 +1,109 @@
+
+import "./MockTetuStrategy.sol";
+import "../relayer/special/TetuShareValueHelper.sol";
+
+contract MockTetuSmartVault is ITetuSmartVault, TestToken, TetuShareValueHelper {
+    using SafeERC20 for IERC20;
+    using FixedPoint for uint256;
+
+    IERC20 public underlyingAsset;
+    uint256 private immutable _underlyingDecimals;
+    uint256 private _underlyingBalanceInVault = 0;
+    MockTetuStrategy private immutable _tetuStrategy;
+    uint256 _desiredRate;
+
+    constructor(
+        string memory name,
+        string memory symbol,
+        uint8 decimals,
+        address _underlyingAsset,
+        MockTetuStrategy tetuStrategy
+    ) TestToken(name, symbol, decimals) {
+        underlyingAsset = IERC20(_underlyingAsset);
+        _underlyingDecimals = decimals;
+        _tetuStrategy = tetuStrategy;
+    }
+
+    function getPricePerFullShare() external pure override returns (uint256) {
+        revert("Should not call this");
+    }
+
+    // Should pass rate with decimals from underlyingToken
+    function setRate(uint256 newRate) public {
+        // stores latest rate, so the balances are recalculated after deposit and withdraw
+        _desiredRate = newRate;
+        uint256 totalSupply = this.totalSupply();
+        // arbitrary number, just to make sure that both Vault and Invested values compose the rate.
+        uint8 vaultInvestedRatio = 3;
+        uint256 totalBalance = (newRate * totalSupply) / 10**_underlyingDecimals;
+        _underlyingBalanceInVault = totalBalance / vaultInvestedRatio;
+        _tetuStrategy.setInvestedUnderlyingBalance(totalBalance - _underlyingBalanceInVault);
+    }
+
+    function underlyingBalanceInVault() external view override returns (uint256) {
+        return _underlyingBalanceInVault;
+    }
+
+    function underlyingBalanceWithInvestmentForHolder(address) external view override returns (uint256) {
+        return underlyingAsset.balanceOf(address(this));
+    }
+
+    function deposit(uint256 amount) external override {
+        IERC20(underlyingAsset).safeTransferFrom(msg.sender, address(this), amount);
+        // Makes sure rate balances are correctly set before calculating wrappedAmount
+        setRate(_desiredRate);
+        uint256 wrappedAmount = _toTetuAmount(amount, this);
+        _mint(msg.sender, wrappedAmount);
+        // Since rate calculation depends on totalSupply, we need to recalculate parameters
+        // that are base to rate calculation.
+        setRate(_desiredRate);
+    }
+
+    function depositFor(uint256, address) external pure override {
+        revert("Should not call this");
+    }
+
+    function withdraw(uint256 numberOfShares) external override {
+        _burn(msg.sender, numberOfShares);
+        // Since rate calculation depends on totalSupply, we need to recalculate parameters
+        // that are base to rate calculation.
+        setRate(_desiredRate);
+        uint256 mainAmount = _fromTetuAmount(numberOfShares, this);
```

Comment:
I think we should calculate this before we burn shares, or else it will be the wrong amount, right?

Reply:
Not really. setRate makes sure the rate is the same, by recalculating the balances of the protocol based on total supply. I'd need to run "setRate" anyway, since tokens can be minted outside of this contract.

Reply:
Ah right, I keep forgetting that the real balance of underlying tokens doesn't matter here. `setRate()` changes the internal accounting to match the supply.